v-next/hardhat/src/internal/builtin-plugins/solidity/build-system/compilation-job.ts

```diff
@@ -102,20 +102,41 @@ export class CompilationJobImplementation implements CompilationJob {
     const defaultOutputSelection: CompilerInput["settings"]["outputSelection"] =
       {
         "*": {
+          "": ["ast"],
           "*": [
             "abi",
             "evm.bytecode",
             "evm.deployedBytecode",
             "evm.methodIdentifiers",
             "metadata",
           ],
-          "": ["ast"],
         },
       };
 
-    // TODO: Deep merge the user output selection with the default one
     const outputSelection = defaultOutputSelection;
 
+    const configOutputSelection = settings.outputSelection ?? {};
+
+    for (const fileKey of Object.keys(configOutputSelection).sort()) {
+      const configFileOutputSelection = configOutputSelection[fileKey] ?? {};
+      outputSelection[fileKey] ??= {};
+
+      for (const contractKey of Object.keys(configFileOutputSelection).sort()) {
```

Comment:
I don't think sorting the keys is strictly necessary because they order of keys isn't expected to change, but it's better to keep the object in a normalized state.

Reply:
Maybe not, actually.  I can't quite contrive a situation in which a plugin would reasonably change key order without adding or removing keys.